### PR TITLE
Feature/char model

### LINKED ISSUE #9: Simple character models

### PR DESCRIPTION
**Pull request by EmilSchier**

### CODE REVIEW

File: app.js [hunk 1]
```diff
@@ -7,11 +7,12 @@ const session = require('express-session');
 var indexRouter = require('./routes/index');
 var usersRouter = require('./routes/users');
 var auth = require('./routes/auth');
+var charactersRouter = require('./routes/characters')
 
 const app = express();
 
 mongoose.Promise = global.Promise;
-mongoose.connect('mongodb://localhost/test', { useNewUrlParser: true, useUnifiedTopology: true, useFindAndModify: false })
+mongoose.connect('mongodb+srv://EmilSchier:Minkodehertiler1%21@node-rest-shop-u8mvl.mongodb.net/test?retryWrites=true&w=majority', { useNewUrlParser: true, useUnifiedTopology: true, useFindAndModify: false })
```
Comment (mschier) [inline app.js]: We need to put these environment variables in a separate file, that is ignored by git. That way we can each have have the secrets ;)

File: models/Character.js [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+    type: [Number],
+    default:  [15, 14, 13, 12, 10, 8]
+  },
+  race: {type: mongoose.Schema.Types.ObjectId, ref: 'Race'}, //, required: true},
+  class: {type: mongoose.Schema.Types.ObjectId, ref: 'CharacterClass'} //, required: true}
+});
+
+//CharacterSchema.statics.findOrCreate = require("find-or-create");
+
+module.exports = mongoose.model('Character', CharacterSchema);
```
Comment (mschier) [inline models/Character.js]: Always end files with a newline :)

File: models/CharacterClass.js [hunk 3]
```diff
@@ -0,0 +1,18 @@
...
+var CharacterClassSchema = new mongoose.Schema({
+  _id: mongoose.Schema.Types.ObjectId,
+  name: String,
+  speed: Number,
+  description: String,
+  savingThrows: {
+    type: [String],
+    default: undefined
+  },
+  hitDie: String
```
Comment (mschier) [inline models/CharacterClass.js]: Can we add some schema validation on the format here such as an enum?
An extra line like `enum: ['d4', 'd6', 'd8', 'd10', 'd12'],` should enforce that only those values can be added.
See [this](https://mongoosejs.com/docs/validation.html#built-in-validators).

File: models/CharacterClass.js [hunk 4]
```diff
@@ -0,0 +1,18 @@
+var mongoose = require('mongoose');
+
+var CharacterClassSchema = new mongoose.Schema({
+  _id: mongoose.Schema.Types.ObjectId,
+  name: String,
+  speed: Number,
+  description: String,
+  savingThrows: {
+    type: [String],
```
Comment (mschier) [inline models/CharacterClass.js]: Consider adding an enum here too (sorry, wrote the comments backwards).

File: models/Character.js [hunk 5]
```diff
@@ -0,0 +1,23 @@
+var mongoose = require('mongoose');
+//var Race = require('../models/Race')
+
+var CharacterSchema = new mongoose.Schema({
+  _id: mongoose.Schema.Types.ObjectId,
+  name: String,
+  level: {type: Number, default: 1},
```
Comment (mschier) [inline models/Character.js]: Consider adding more restrictions like 'min' and 'max' :)

File: routes/characters.js [hunk 6]
```diff
@@ -0,0 +1,134 @@
...
+        }
+      });
+    })
+    .catch(err => {
+      console.log(err);
+      res.status(500).json({ error: err });
+    });
+});
+
+router.get('/:charactertId', (req, res, next) => {
```
Comment (mschier) [inline routes/characters.js]: An extra t as found it's way into the word 'charactertId' ^^

File: routes/characters.js [hunk 7]
```diff
@@ -0,0 +1,134 @@
...
+  const id = req.params.charactertId;
+  Character.findById(id)
+    .select('_id name level proficiencyBonus hp hpMax speed description abilityScores race class')
```
Comment (mschier) [inline routes/characters.js]: Is it necessary to write all fields, when we want everything? isn't there a shorthand for selecting everything?

File: models/Character.js [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+  hp: Number,
+  hpMax: Number,
+  speed: Number,
+  description: String,
+  abilityScores: {
+    type: [Number],
+    default:  [15, 14, 13, 12, 10, 8]
+  },
+  race: {type: mongoose.Schema.Types.ObjectId, ref: 'Race'}, //, required: true},
+  class: {type: mongoose.Schema.Types.ObjectId, ref: 'CharacterClass'} //, required: true}
```
Comment (mschier) [inline models/Character.js]: We should be consistent in naming - if we decide that characterClass is the term we'll use, we should name the field name in the same way :)

File: models/CharacterClass.js [hunk 9]
```diff
@@ -0,0 +1,18 @@
...
+    default: undefined
+  },
+  hitDie: String
+  
+});
+
+//RaceSchema.statics.findOrCreate = require("find-or-create");
+
+module.exports = mongoose.model('CharacterClass', CharacterClassSchema);
```
Comment (mschier) [inline models/CharacterClass.js]: Newline again ^^

File: models/Race.js [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+  abilityScoreModifires: {
+    type: [Number],
+    default:  [0, 0, 0, 0, 0, 0]
+  }
+  
+});
+
+//RaceSchema.statics.findOrCreate = require("find-or-create");
+
+module.exports = mongoose.model('Race', RaceSchema);
```
Comment (mschier) [inline models/Race.js]: Newline again ^^

File: routes/characters.js [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+      });
+    })
+    .catch(err => {
+      console.log(err);
+      res.status(500).json({ error: err });
+    });
+});
+
+
+module.exports = router;
```
Comment (mschier) [inline routes/characters.js]: Newline again ^^

File: models/CharacterClass.js [hunk 12]
```diff
@@ -0,0 +1,18 @@
+var mongoose = require('mongoose');
+
+var CharacterClassSchema = new mongoose.Schema({
+  _id: mongoose.Schema.Types.ObjectId,
+  name: String,
+  speed: Number,
```
Comment (l-schier) [inline models/CharacterClass.js]: Classes har ikke speeds

File: models/Character.js [hunk 13]
```diff
@@ -0,0 +1,23 @@
...
+//var Race = require('../models/Race')
+
+var CharacterSchema = new mongoose.Schema({
+  _id: mongoose.Schema.Types.ObjectId,
+  name: String,
+  level: {type: Number, default: 1},
+  proficiencyBonus: Number,
+  hp: Number,
+  hpMax: Number,
+  speed: Number,
```
Comment (l-schier) [inline models/Character.js]: Speed kommer reelt fra racen jo

File: README.md [hunk 14]
```diff
@@ -1,2 +1,29 @@
...
+	 "level": "10",
+    "abilityScores": [
+        "20",
+        "20",
+        "20",
+        "20",
+        "20",
+        "20"
+    ],
+    "_id": "5e42c05e0b7a2c3cc094bc90",
```
Comment (mschier) [inline README.md]: _id is not needed, when POSTing a new character, it will be generated by Mongo/Mongoose ;)

**Review by l-schier — APPROVED** [review]

File: README.md [hunk 15]
```diff
@@ -1,2 +1,29 @@
...
+    "name": "Test",
+    "proficiencyBonus": "5",
+    "hp": "43",
+    "hpMax": "43",
+    "speed": "30",
+    "description": "Endnu en test før jeg oploader det"
+	
+}
+ POST Request to http://localhost:3000/characters/characterId (if on local host) with a Json body similar to seen bellow will update the given parameters of the Character with characterId.
```
Comment (mschier) [inline README.md]: Suggested: `PATCH request to http://localhost:3000/characters/{characterId}`

File: models/Character.js (hunk 5 above)
Comment (mschier) [reply models/Character.js]: Consider adding `min: 0` for all only positive numbers :)

**Review by mschier — CHANGES_REQUESTED** [review]
In general very nice! :D I think there's a bit more could be done with validations in the Mongoose models, and we're missing quite a few terminating newlines. ^^

File: routes/characters.js (hunk 7 above)
Comment (l-schier) [reply routes/characters.js]: image
Like this in the mongoDB documentation

File: README.md (hunk 14 above)
Comment (EmilSchier) [reply README.md]: true, this went quite fast

File: routes/characters.js (hunk 7 above)
Comment (EmilSchier) [reply routes/characters.js]: There is a field __v in the object, that i dont really want to display, is there a way to just specify what i dont want to see?

File: app.js (hunk 1 above)
Comment (EmilSchier) [reply app.js]: I must admit i do nok quite know how to do this

File: README.md (hunk 15 above)
Comment (mschier) [reply README.md]: Additionally, we can just specify `PATCH /characters/{characterId}` :)

File: app.js (hunk 1 above)
Comment (mschier) [reply app.js]: It's okay, I'll make some changes in another branch :)

**Review by l-schier — APPROVED** [review]
Looks very good with the fixes on some of the earlier problems

File: README.md [hunk 16]
```diff
@@ -1,17 +1,16 @@
 # DnDeep-backend
 The backend part of DnDeep
 
-to create a Character send a POST request to http://localhost:3000/characters/ (if on local host) with a body similar to what is seen bellow:
+to create a Character send a POST request to http://{localhost}:3000/characters/ (if on local host) with a body similar to what is seen bellow:
```
Comment (l-schier) [inline README.md]: This is my own mistake, codacy wants to have angle brackets on each side like this < Link > on all literal urls

**Review by l-schier — APPROVED** [review]

File: README.md (hunk 16 above)
Comment (l-schier) [reply README.md]: I missunderstood and so made it curly brackets instead which should probably be removed

Comment (mschier) [reply README.md]: I've pushed a new commit with better Markdown :)

File: models/CharacterClass.js (hunk 12 above)
Comment (mschier) [reply models/CharacterClass.js]: Already fixed :)

File: models/Character.js (hunk 13 above)
Comment (mschier) [reply models/Character.js]: Already fixed :)

File: routes/characters.js (hunk 7 above)
Comment (mschier) [reply routes/characters.js]: I see that you figured it out by adding `.select('-__v')` instead 👍

File: models/Character.js (hunk 8 above)
Comment (mschier) [reply models/Character.js]: Already fixed :)

File: models/Character.js [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+  race: {
+    type: mongoose.Schema.Types.ObjectId,
+    ref: 'Race'//,
+//  required: true  
+  }, 
+  characterClass: {
+    type: mongoose.Schema.Types.ObjectId,
+    ref: 'CharacterClass'//,
+//  required: true
+  }
```
Comment (mschier) [inline models/Character.js:45]: This should be a list of CharacterClasses, to fit our needs?

**Review by mschier — APPROVED** [review]

File: models/Character.js (hunk 17 above)
Comment (mschier) [reply models/Character.js:45]: @l-schier and @EmilSchier, what do You guys think?

Comment (l-schier) [reply models/Character.js:45]: It should be a list yes, however we need to revise the way we handle the level on top of that and figure out how we manage multiclassing

**Review by l-schier — APPROVED** [review]
Good fixes

File: models/Character.js (hunk 17 above)
Comment (mschier) [reply models/Character.js:45]: I agree, let's rethink it later :)

**Pull request merged**